File: src/mruby_integration/shapes/line.cpp
#include "mruby.h"
#include "mruby/array.h"
#include "raylib.h"

#include "mruby_integration/struct_types.hpp"

auto
mrb_draw_line(mrb_state* mrb, mrb_value) -> mrb_value
{
  mrb_int start_x, start_y, end_x, end_y;
  Color* colour;

  mrb_get_args(
    mrb, "iiiid", &start_x, &start_y, &end_x, &end_y, &colour, &Colour_type);

  DrawLine(start_x, start_y, end_x, end_y, *colour);
  return mrb_nil_value();
}

auto
mrb_draw_line_v(mrb_state* mrb, mrb_value) -> mrb_value
{
  Vector2 *start, *end;
  Color* colour;

  mrb_get_args(mrb,
               "ddd",
               &start,
               &Vector2_type,
               &end,
               &Vector2_type,
               &colour,
               &Colour_type);

  DrawLineV(*start, *end, *colour);
  return mrb_nil_value();
}

auto
mrb_draw_line_ex(mrb_state* mrb, mrb_value) -> mrb_value
{
  Vector2 *start, *end;
  mrb_float thickness;
  Color* colour;

  mrb_get_args(mrb,
               "ddfd",
               &start,
               &Vector2_type,
               &end,
               &Vector2_type,
               &thickness,
               &colour,
               &Colour_type);

  DrawLineEx(*start, *end, thickness, *colour);
  return mrb_nil_value();
}

auto
mrb_draw_line_bezier(mrb_state* mrb, mrb_value) -> mrb_value
{
  Vector2 *start, *end;
  mrb_float thickness;
  Color* colour;

  mrb_get_args(mrb,
               "ddfd",
               &start,
               &Vector2_type,
               &end,
               &Vector2_type,
               &thickness,
               &colour,
               &Colour_type);

  DrawLineBezier(*start, *end, thickness, *colour);
  return mrb_nil_value();
}

auto
mrb_draw_line_bezier_quad(mrb_state* mrb, mrb_value) -> mrb_value
{
  Vector2 *start, *end, *control;
  mrb_float thickness;
  Color* colour;

  mrb_get_args(mrb,
               "dddfd",
               &start,
               &Vector2_type,
               &end,
               &Vector2_type,
               &control,
               &Vector2_type,
               &thickness,
               &colour,
               &Colour_type);

  DrawLineBezierQuad(*start, *end, *control, thickness, *colour);
  return mrb_nil_value();
}

auto
mrb_draw_line_strip(mrb_state* mrb, mrb_value) -> mrb_value
{
  mrb_value vectors;
  Color* colour;

  mrb_get_args(mrb, "Ad", &vectors, &colour, &Colour_type);

  mrb_int array_length = RARRAY_LEN(vectors);

  Vector2* raw_vectors[array_length];

  for (mrb_int i = 0; i < array_length; i++) {
    raw_vectors[i] =
      DATA_GET_PTR(mrb, mrb_ary_ref(mrb, vectors, i), &Vector2_type, Vector2);
  }

  DrawLineStrip(*raw_vectors, array_length, *colour);
  return mrb_nil_value();
}

void
append_shapes_line(mrb_state* mrb)
{
  mrb_define_method(
    mrb, mrb->kernel_module, "draw_line", mrb_draw_line, MRB_ARGS_REQ(5));
  mrb_define_method(
    mrb, mrb->kernel_module, "draw_line_v", mrb_draw_line_v, MRB_ARGS_REQ(3));
  mrb_define_method(
    mrb, mrb->kernel_module, "draw_line_ex", mrb_draw_line_ex, MRB_ARGS_REQ(4));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "draw_line_bezier",
                    mrb_draw_line_bezier,
                    MRB_ARGS_REQ(4));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "draw_line_bezier_quad",
                    mrb_draw_line_bezier_quad,
                    MRB_ARGS_REQ(5));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "draw_line_strip",
                    mrb_draw_line_strip,
                    MRB_ARGS_REQ(2));
}

File: mrb_doc/shapes/line.rb
# Draw a line between the start and end points.
# @param start_x [Integer]
# @param start_y [Integer]
# @param end_x [Integer]
# @param end_y [Integer]
# @param colour [Colour]
# @return [nil]
def draw_line(start_x, start_y, end_x, end_y, colour)
  # mrb_draw_line
  # src/mruby_integration/shapes/line.cpp
  nil
end

# Draw a line between the start and end points using {Vector2}s.
# @param start [Vector2]
# @param stop [Vector2]
# @param colour [Colour]
# @return [nil]
def draw_line_v(start, stop, colour)
  # mrb_draw_line_v
  # src/mruby_integration/shapes/line.cpp
  nil
end

# Draw a line between the start and end points using {Vector2}s and specified thickness.
# @param start [Vector2]
# @param stop [Vector2]
# @param thickness [Float]
# @param colour [Colour]
# @return [nil]
def draw_line_ex(start, stop, thickness, colour)
  # mrb_draw_line_ex
  # src/mruby_integration/shapes/line.cpp
  nil
end

# Draw a bezier curve between the start and end points using {Vector2}s and specified thickness.
# @param start [Vector2]
# @param stop [Vector2]
# @param thickness [Float]
# @param colour [Colour]
# @return [nil]
def draw_line_bezier(start, stop, thickness, colour)
  # mrb_draw_line_bezier
  # src/mruby_integration/shapes/line.cpp
  nil
end

# Draw a bezier curve between the start and end points passing through the control using {Vector2}s and specified thickness
# @param start [Vector2]
# @param stop [Vector2]
# @param control [Vector2]
# @param thickness [Float]
# @param colour [Colour]
# @return [nil]
def draw_line_bezier_quad(start, stop, control, thickness, colour)
  # mrb_draw_line_bezier_quad
  # src/mruby_integration/shapes/line.cpp
  nil
end

# @note I honestly don't know how this method is supposed to work, I thought I had
#   bugs in my code but reproducing it in plain C++ causes it to do the same
#   things. As far as I can tell it just draws a line from 0, 0 to the first
#   vector and nothing else.
# @param vectors [Array<Vector2>]
# @param colour [Colour]
# @return [nil]
def draw_line_strip(vectors, colour)
  # mrb_draw_line_strips
  # src/mruby_integration/shapes/line.cpp
  nil
end
